iroha_data_model/src/lib.rs

```diff
@@ -133,157 +134,112 @@ impl Value {
+                        Ok(id)
+                    } else {
+                        Err(Self::Error::default())
+                    }
+                }
+            }
 
-impl From<u32> for Value {
-    fn from(value: u32) -> Value {
-        Value::U32(value)
-    }
+            impl From<$ty> for Value {
+                fn from(id: $ty) -> Self {
+                    Value::Id(IdBox::$variant(id))
+                }
+            }
+        )*
+    };
 }
 
-impl From<bool> for Value {
-    fn from(value: bool) -> Value {
-        Value::Bool(value)
-    }
-}
+from_and_try_from_value_idbox!(
+    AccountId(account::Id),
+    AssetId(asset::Id),
+    AssetDefinitionId(asset::DefinitionId),
+    PeerId(peer::Id),
+);
+// TODO: Should we wrap String with new type in order to convert like here?
+//from_and_try_from_value_idbox!((DomainName(Name), ErrorValueTryFromDomainName),);
+
+macro_rules! from_and_try_from_value_identifiablebox {
+    ( $(box $variant:ident( $ty:ty ),)* ) => {
+        $(
+            impl TryFrom<Value> for $ty {
+                type Error = ErrorTryFromEnum<Self, Value>;
+
+                fn try_from(value: Value) -> Result<Self, Self::Error> {
+                    if let Value::Identifiable(IdentifiableBox::$variant(id)) = value {
+                        Ok(*id)
+                    } else {
+                        Err(Self::Error::default())
+                    }
+                }
+            }
 
-impl From<Parameter> for Value {
-    fn from(value: Parameter) -> Value {
-        Value::Parameter(value)
-    }
-}
+            impl From<$ty> for Value {
+                fn from(id: $ty) -> Self {
+                    Value::Identifiable(IdentifiableBox::$variant(Box::new(id)))
+                }
+            }
+        )*
+    };
+    ( $( $variant:ident( $ty:ty ), )* ) => {
+        $(
+            impl TryFrom<Value> for $ty {
+                type Error = ErrorTryFromEnum<Self, Value>;
+
+                fn try_from(value: Value) -> Result<Self, Self::Error> {
+                    if let Value::Identifiable(IdentifiableBox::$variant(id)) = value {
+                        Ok(id)
+                    } else {
+                        Err(Self::Error::default())
+                    }
+                }
+            }
 
-impl From<IdentifiableBox> for Value {
-    fn from(value: IdentifiableBox) -> Value {
-        Value::Identifiable(value)
-    }
+            impl From<$ty> for Value {
+                fn from(id: $ty) -> Self {
+                    Value::Identifiable(IdentifiableBox::$variant(id))
+                }
+            }
+        )*
+    };
 }
 
-impl From<IdBox> for Value {
-    fn from(value: IdBox) -> Value {
-        Value::Id(value)
-    }
-}
+from_and_try_from_value_identifiablebox!(
+    box Account(account::Account),
+    box Asset(asset::Asset),
+    box AssetDefinition(asset::AssetDefinition),
+    box Domain(domain::Domain),
+    box Peer(peer::Peer),
+);
```

Comment:
As far as I know the box syntax is unstable, is this custom syntax?

Reply:
https://github.com/rust-lang/rust/issues/49733

Reply:
It is just a syntax word for pattern matching inside macros

Reply:
Then why do we have two `from_and_try_from_value_identifiablebox!` implementations here?

Reply:
Because there are 2 cases. 

1 case is `impl From<Box<Var>> for Enum`
2 case is `impl From<Var> for Enum`

Reply:
Ok, I see, this should be separated into 2 macros to my mind. This will make it a lot easier to understand.

Reply:
And we can remove the custom syntax `box` in this case.